File: VB/App_Code/SessionDashboardStorage.vb
Imports System.Web.SessionState
Imports DevExpress.DashboardWeb

Namespace DevExpress.PublicDemo
    Public Class SessionDashboardStorage
        Inherits DashboardStorageBase

        Private Const DashboardStorageKey As String = "DashboardStorage"


        Private Shared instance_Renamed As SessionDashboardStorage = Nothing

        Public Shared ReadOnly Property Instance() As SessionDashboardStorage
            Get
                If instance_Renamed Is Nothing Then
                    instance_Renamed = New SessionDashboardStorage()
                End If
                Return instance_Renamed
            End Get
        End Property

        Private ReadOnly Property Storage() As Dictionary(Of String, XDocument)
            Get
                Dim session As HttpSessionState = HttpContext.Current.Session
                If session IsNot Nothing Then

                    Dim storage_Renamed As Dictionary(Of String, XDocument) = TryCast(session(DashboardStorageKey), Dictionary(Of String, XDocument))
                    If storage_Renamed Is Nothing Then
                        storage_Renamed = New Dictionary(Of String, XDocument)()
                        session(DashboardStorageKey) = storage_Renamed
                        Return storage_Renamed
                    End If
                    Return storage_Renamed
                End If
                Throw New ArgumentNullException("Session is not available")
            End Get
        End Property

        Protected Sub New()
            MyBase.New()

        End Sub

        Protected Overrides Function GetAvailableDashboardsID() As IEnumerable(Of String)
            Return Storage.Keys
        End Function
        Protected Overrides Function LoadDashboard(ByVal dashboardID As String) As XDocument
            Return Storage(dashboardID)
        End Function
        Protected Overrides Sub SaveDashboard(ByVal dashboardID As String, ByVal dashboard As XDocument, ByVal createNew As Boolean)
            If createNew Xor Storage.ContainsKey(dashboardID) Then
                Storage(dashboardID) = dashboard
            End If
        End Sub
        Public Sub RegisterDashboard(ByVal dashboardID As String, ByVal dashboard As XDocument)
            SaveDashboard(dashboardID, dashboard, True)
        End Sub
    End Class
End Namespace

File: CS/App_Code/SessionDashboardStorage.cs
using System;
using System.Collections.Generic;
using System.Xml.Linq;
using DevExpress.DashboardCommon;
using DevExpress.DashboardWeb;
using System.Web;
using System.Web.SessionState;

namespace DevExpress.PublicDemo {
    public class SessionDashboardStorage : DashboardStorageBase {
        const string DashboardStorageKey = "DashboardStorage";

        private static SessionDashboardStorage instance = null;

        public static SessionDashboardStorage Instance {
            get {
                if(instance == null) {
                    instance = new SessionDashboardStorage();
                }
                return instance;
            }
        }

        Dictionary<string, XDocument> Storage {
            get {
                HttpSessionState session = HttpContext.Current.Session;
                if(session != null) {
                    Dictionary<string, XDocument> storage = session[DashboardStorageKey] as Dictionary<string, XDocument>;
                    if(storage == null) {
                        storage = new Dictionary<string, XDocument>();
                        session[DashboardStorageKey] = storage;
                        return storage;
                    }
                    return storage;
                }
                throw new ArgumentNullException("Session is not available");
            }
        }

        protected SessionDashboardStorage() : base() {

        }

        protected override IEnumerable<string> GetAvailableDashboardsID() {
            return Storage.Keys;
        }
        protected override XDocument LoadDashboard(string dashboardID) {
            return Storage[dashboardID];
        }
        protected override void SaveDashboard(string dashboardID, XDocument dashboard, bool createNew) {
            if(createNew ^ Storage.ContainsKey(dashboardID))
                Storage[dashboardID] = dashboard;
        }
        public void RegisterDashboard(string dashboardID, XDocument dashboard) {
            SaveDashboard(dashboardID, dashboard, true);
        }
    }
}